src/main/scala/mesosphere/marathon/core/storage/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -0,0 +1,142 @@
+    ctx: ExecutionContext) extends PersistenceStore[K, Category, Serialized] with StrictLogging {
+
+  private[storage] val idCache = RwLock(TrieMap[Category, Seq[Any]]())
+  private[storage] val valueCache = RwLock(TrieMap[K, Option[Any]]())
+
+  override def ids[Id, V]()(implicit ir: IdResolver[Id, V, Category, K]): Source[Id, NotUsed] = {
+    val storeId = ir.category
+    idCache.read { ids =>
+      if (ids.contains(storeId)) {
+        Source(ids(storeId).map(_.asInstanceOf[Id]))
+      } else {
+        val childFuture = async {
+          val children = await(store.ids.toMat(Sink.seq)(Keep.right).run())
+          idCache.write(c => c(storeId) = children)
+          children
+        }
+        Source.fromFuture(childFuture).mapConcat(identity)
+      }
+    }
+  }
+
+  override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    async {
+      await(store.deleteAll(k))
+      val storageId = ir.toStorageId(k, None)
+      valueCache.write { vc =>
+        vc.remove(storageId)
+        idCache.write { c =>
+          val old = c.getOrElse(ir.category, Nil)
+          val children = old.filter(_ != storageId)
+          if (children.nonEmpty) {
+            c.put(ir.category, children)
+          } else {
+            c.remove(ir.category)
+          }
+        }
+      }
+      Done
+    }
+  }
+
+  override def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = {
+    val storageId = ir.toStorageId(id, None)
+    val cached = valueCache.read(_.get(storageId))
+    cached match {
+      case Some(v) =>
+        Future.successful(v.asInstanceOf[Option[V]])
+      case None =>
+        async {
+          val value = await(store.get(id))
+          valueCache.write(_.put(storageId, value))
+          value
+        }
+    }
+  }
+
+  override def get[Id, V](id: Id, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] =
+    store.get(id, version)
+
+  override def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done] = {
+    async {
+      await(store.store(id, v))
+      val storageId = ir.toStorageId(id, None)
+      val storageCategory = ir.category
+      valueCache.write { values =>
+        idCache.write { ids =>
+          values.put(storageId, Some(v))
+          val cachedIds = ids.getOrElse(storageCategory, Nil)
+          ids.put(storageCategory, id +: cachedIds)
+        }
+      }
+      Done
+    }
+  }
+
+  override def store[Id, V](id: Id, v: V, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized]): Future[Done] = async {
+    await(store.store(id, v, version))
+    valueCache.write { vc =>
+      vc.putIfAbsent(ir.toStorageId(id, None), Some(v))
```

Comment:
It is actually intended. store(id, v, version) says "if there is no current, make it current". I'll add a note to that effect.